### PR TITLE
New attributes and methods for stratigraphic layers

### PR DESCRIPTION
**Pull request by SergeyTsimfer**

### CODE REVIEW

File: seismiqb/labels/horizon/attributes.py [hunk 1]
```diff
@@ -189,14 +190,22 @@ def binary_matrix(self):
...
         """ An alias. """
         return self.full_binary_matrix
 
+    @cached_property
+    def float_matrix(self):
+        """ Matrix with smoothed float values in cubic (ordinal) coordinates. """
+        if self.dtype == np.float32:
+            return self.full_matrix
+        smoothed = self.smooth_out(mode='convolve', kernel_size=5, sigma_spatial=3, max_depth_difference=5,
+                                    inplace=False, dtype=np.float32)
```
Comment (HollowPrincess) [inline seismiqb/labels/horizon/attributes.py:207]: Extra space on the second row

File: seismiqb/labels/horizon/attributes.py [hunk 2]
```diff
@@ -344,78 +359,102 @@ def get_cube_values(self, window=1, offset=0, chunk_size=256, src_geometry=None,
...
+        result[~self.full_binary_matrix] = np.nan
 
+        if apply_float_correction:
+            result = self._apply_float_correction(result)
+        return result
 
-    def get_array_values(self, array, shifts=None, grid_info=None, width=5, axes=(2, 1, 0)):
-        """ Get values from an external array along the horizon.
+    def get_layer_values(self, other, window=0, offset=0, src_geometry=None):
+        """ Get values from the cube between `self` and `other` horizons.
```
Comment (HollowPrincess) [inline seismiqb/labels/horizon/attributes.py:378]: Is this function more about horizons than the field? Looks like this is not a suitable place for it.

File: seismiqb/labels/horizon/attributes.py [hunk 3]
```diff
@@ -344,78 +359,102 @@ def get_cube_values(self, window=1, offset=0, chunk_size=256, src_geometry=None,
...
+
+        self_depths  -= d_start + window[1]
+        other_depths -= d_start - window[0]
+
+        # Load data chunk, prepare output array
+        data = geometry[:, :, d_start:d_stop]
+        n = d_stop - d_start
+        result = np.full((*presence_matrix.shape, n), fill_value=np.nan, dtype=np.float32)
+
+        # Filter unnecessary points, copy amplitudes from `data`, shift `self_depth` indices
```
Comment (HollowPrincess) [inline seismiqb/labels/horizon/attributes.py:413]: The comment doesn't make the situation clearer

File: seismiqb/labels/horizon/attributes.py [hunk 4]
```diff
@@ -344,78 +359,102 @@ def get_cube_values(self, window=1, offset=0, chunk_size=256, src_geometry=None,
...
+            other_depths = other_depths[mask]
+
+            result[idx_i, idx_x, i] = data[idx_i, idx_x, self_depths]
+            self_depths += 1
+
+        result[~presence_matrix] = np.nan
         return result
 
+    def _apply_float_correction(self, array):
+        """ Compute a three-point correction from int to float.
```
Comment (HollowPrincess) [inline seismiqb/labels/horizon/attributes.py:428]: Compute or apply?

File: seismiqb/labels/horizon/attributes.py [hunk 5]
```diff
@@ -344,78 +359,102 @@ def get_cube_values(self, window=1, offset=0, chunk_size=256, src_geometry=None,
...
+        As most of the loaded horizons are loaded in int32 values and only integer samples can be retrieved from
```
Comment (HollowPrincess) [inline seismiqb/labels/horizon/attributes.py:429]: **loaded** horizons are **loaded**

File: seismiqb/labels/horizon/attributes.py [hunk 6]
```diff
@@ -344,78 +359,102 @@ def get_cube_values(self, window=1, offset=0, chunk_size=256, src_geometry=None,
...
+        the cube, extracted values (amplitudes) and their derivatives may have step-like artifacts.
+        To eliminate them, we interpolate between depth-wise channels, using the size of difference between
+        int32 and float32 horizons as the weight for interpolation.
+
+        Parameters
+        ----------
+        array : array-like
+            Array of (I, X, C) shape.
```
Comment (HollowPrincess) [inline seismiqb/labels/horizon/attributes.py:437]: Why C? Maybe D?

File: seismiqb/labels/horizon/attributes.py [hunk 7]
```diff
@@ -344,78 +359,102 @@ def get_cube_values(self, window=1, offset=0, chunk_size=256, src_geometry=None,
...
+        -------
+        Array of (I, X, C-2) shape.
+        """
+        shifts_matrix = self.float_matrix - self.full_matrix
+        middle = array[:, :, 1:-1]
+        output = middle.copy()
+
+        mask = shifts_matrix > 0
+        output[mask] = (1 - shifts_matrix[mask]).reshape(-1, 1) * middle[mask] + \
+                        +(shifts_matrix[mask]).reshape(-1, 1) * array[:, :, 2:][mask]
```
Comment (HollowPrincess) [inline seismiqb/labels/horizon/attributes.py:449]:
```suggestion
                        + (shifts_matrix[mask]).reshape(-1, 1) * array[:, :, 2:][mask]
```

File: seismiqb/labels/horizon/attributes.py [hunk 8]
```diff
@@ -344,78 +359,102 @@ def get_cube_values(self, window=1, offset=0, chunk_size=256, src_geometry=None,
...
+
+        mask = shifts_matrix < 0
+        output[mask] = (1 + shifts_matrix[mask]).reshape(-1, 1) * middle[mask] + \
+                        +(-shifts_matrix[mask]).reshape(-1, 1) * array[:, :, :-2][mask]
```
Comment (HollowPrincess) [inline seismiqb/labels/horizon/attributes.py:453]:
```suggestion
                        + (-shifts_matrix[mask]).reshape(-1, 1) * array[:, :, :-2][mask]
```

File: seismiqb/labels/horizon/attributes.py (hunk 1 above)
Comment (HollowPrincess) [reply seismiqb/labels/horizon/attributes.py:207]: Why is it a full matrix?

File: seismiqb/labels/horizon/attributes.py [hunk 9]
```diff
@@ -501,6 +540,13 @@ def load_attribute(self, src, location=None, use_cache=True, enlarge=False, **kw
             src_name = src.pop('src')
             kwargs.update(src)
 
+        if '[' in src_name:
+            pos_, _pos = src_name.find('['), src_name.find(']')
+            channels = literal_eval(src_name[1+pos_:_pos])
+            src_name = src_name[:pos_]
+        else:
+            channels = None
+
```
Comment (HollowPrincess) [inline seismiqb/labels/horizon/attributes.py:549]: Why do we need this?

File: seismiqb/labels/horizon/attributes.py (hunk 1 above)
Comment (SergeyTsimfer) [reply seismiqb/labels/horizon/attributes.py:207]: I was thinking about introducing both `float_matrix` and `full_float_matrix`, but decided against it as the only current usage requires `full_float_matrix`

File: seismiqb/labels/horizon/attributes.py (hunk 5 above)
Comment (SergeyTsimfer) [reply seismiqb/labels/horizon/attributes.py:429]: they are indeed

File: seismiqb/labels/horizon/attributes.py (hunk 9 above)
Comment (SergeyTsimfer) [reply seismiqb/labels/horizon/attributes.py:549]: That is convenient for facies visualization

**Pull request merged**